pravega/pravega-operator
# Issue 97: Update operator SDK dependency to 0.2.0 (full refactor)

## Change log description
The [Operator SDK project](https://github.com/operator-framework/operator-sdk) is moving fast and they have recently moved to a new `0.2.0` minor version that brings new features and completely refactors the operator framework. If we don't want to be left behind, we need to keep updating our operator SDK dependency, specially now that the project is still in an early stage and evolves fast.

Changes:

- Update the operator SDK dependency from `0.0.5` to `0.2.0`. Applied the [official migration guide](https://github.com/operator-framework/operator-sdk/blob/master/doc/migration/v0.1.0-migration-guide.md) step by step to migrate the existing operator project to be compliant with the latest SDK.
- Set up the operator testing framework as explained in the [official documentation](https://github.com/operator-framework/operator-sdk/blob/master/doc/test-framework/writing-e2e-tests.md). There's no test case implementation yet, but the framework is ready to accept test cases.
- Create new `test-e2e` make command that runs the end-to-end tests
- Update Travis configuration to set up and deploy MiniKube as the testing Kubernetes backend.
- The vast majority of the files updated in this PR belong to the `vendor` directory. Ignore those when reviewing the changes.

## Purpose of the change
- Fix #97 

## How to test the change

Checkout the repository (`master` branch is fine, no need to checkout this branch).

Edit the operator image used in the `deploy/operator.yaml` file to use the `adrianmo/pravega-operator:issue-97-3` image, which contains the changes in this PR.

Follow the regular README file to install the operator and deploy the operator.

When you deploy the operator, check the logs to make sure you are using an operator based on the `0.2.0` SDK.

```
$ kubectl logs -f pravega-operator-574bbf8888-v6bht      
INFO[0000] pravega-operator Version: 0.1.2-15           
INFO[0000] Git SHA: 7897372                             
INFO[0000] Go Version: go1.11                           
INFO[0000] Go OS/Arch: linux/amd64                      
INFO[0000] operator-sdk Version: v0.2.0 <-----
```

When you deploy a Pravega cluster, make sure it creates all resources as usual.

```
$ kubectl get all -l pravega_cluster=pravega
NAME                                             READY     STATUS    RESTARTS   AGE                                                                            
pod/pravega-bookie-0                             1/1       Running   0          28m                                                                            
pod/pravega-bookie-1                             1/1       Running   0          28m                                                                            
pod/pravega-bookie-2                             1/1       Running   0          28m                                                                            
pod/pravega-pravega-controller-c9875fc6d-l7prt   1/1       Running   0          28m                                                                            
pod/pravega-segmentstore-0                       1/1       Running   1          28m                                                                            
pod/pravega-segmentstore-1                       1/1       Running   1          28m                                                                            
pod/pravega-segmentstore-2                       1/1       Running   1          28m                                                                            

NAME                                 TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)              AGE                                                        
service/pravega-bookie-headless      ClusterIP   None            <none>        3181/TCP             28m                                                        
service/pravega-pravega-controller   ClusterIP   10.71.254.167   <none>        10080/TCP,9090/TCP   28m                                                        

NAME                                               DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE                                                            
deployment.extensions/pravega-pravega-controller   1         1         1            1           28m                                                            

NAME                                                         DESIRED   CURRENT   READY     AGE                                                                 
replicaset.extensions/pravega-pravega-controller-c9875fc6d   1         1         1         28m                                                                 

NAME                                         DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE                                                                  
deployment.apps/pravega-pravega-controller   1         1         1            1           28m                                                                  

NAME                                                   DESIRED   CURRENT   READY     AGE                                                                       
replicaset.apps/pravega-pravega-controller-c9875fc6d   1         1         1         28m                                                                       

NAME                                    DESIRED   CURRENT   AGE
statefulset.apps/pravega-bookie         3         3         28m
statefulset.apps/pravega-segmentstore   3         3         28m
```
